File: 2022/day2/main.pl
use strict;
use warnings;
use v5.34;

my $ans    = 0;
my $ans2   = 0;
my %wr     = qw(r p p s s r);
my %lr     = qw(r s p r s p);
my %scores = qw(r 1 p 2 s 3);

sub score {
    my $ws = 0;
    $_[0] eq $_[1] ? $ws += 3 : $wr{ $_[0] } eq $_[1] ? $ws += 6 : undef;
    return $scores{ $_[1] } + $ws;
}

sub score2 {
        $_[1] eq "r" ? return $scores{ $lr{ $_[0] } } + 0
      : $_[1] eq "p" ? return $scores{ $_[0] } + 3
      :                return $scores{ $wr{ $_[0] } } + 6;
}

open( my $f, "<", "input.txt" );
while (<$f>) {
    tr/XYZ/ABC/;
    tr/ABC/rps/;
    my ( $c, $s ) = split(/\s/);
    $ans  += score( $c, $s );
    $ans2 += score2( $c, $s );
}
say "Part 1: $ans";
say "Part 2: $ans2";

File: 2022/day2/main.py
ans, ans2 = 0, 0
wr, lr = dict(zip("ABC", "YZX")), dict(zip("ABC", "CAB"))
scores = dict(zip("rps", range(1, 4)))
vr = dict(zip("ABCXYZ", ["r", "p", "s"] * 2))


def score2(c, s):
    if s == "X":
        return scores[vr[lr[c]]] + 0
    elif s == "Y":
        return scores[vr[c]] + 3
    else:
        return scores[vr[wr[c]]] + 6


def score(c, s):
    ws = 0
    if vr[c] == vr[s]:
        ws += 3
    elif wr[c] == s:
        ws += 6
    return ws + scores[vr[s]]


with open("input.txt") as f:
    for line in f:
        parts = line.split()
        c, s = parts[0], parts[1]
        ans += score(c, s)
        ans2 += score2(c, s)

print(f"Part 1: {ans}")
print(f"Part 2: {ans2}")
